### DISCUSSION TITLE
[script-setup] Why not output `defineComponent`?

### DISCUSSION THREAD

**Discussion by wenfangdu** [vuejs/rfcs#375] [RFC Discussions]
[RFC Link](https://github.com/vuejs/rfcs/blob/master/active-rfcs/0040-script-setup.md#basic-example)

Source code:

```vue
<script setup>
  // imported components are also directly usable in template
  import Foo from './Foo.vue'
  import { ref } from 'vue'

  // write Composition API code just like in a normal setup()
  // but no need to manually return everything
  const count = ref(0)
  const inc = () => {
    count.value++
  }
</script>

<template>
  <Foo :count="count" @click="inc" />
</template>
```

Compiled output:

```js
import Foo from './Foo.vue'
import { ref } from 'vue'

//            👇 no `defineComponent` here, why?
export default {
  setup() {
    const count = ref(1)
    const inc = () => {
      count.value++
    }

    return function render() {
      return h(Foo, {
        count,
        onClick: inc,
      })
    }
  },
}
```

**Comment by ShGKme** [comment]
For objects `defineComponent` does nothing in runtime and is useful only for development tools and IDE (useless in generated code).

https://github.com/vuejs/vue-next/blob/4781965cc2396b51028a05917a3b057f884bc0b7/packages/runtime-core/src/apiDefineComponent.ts#L184

**Reply by wenfangdu** [reply]
Adding `lang="ts"` [does output `defineCompoent`](https://sfc.vuejs.org/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8aDE+e3sgbXNnIH19PC9oMT5cbjwvdGVtcGxhdGU+XG5cbjxzY3JpcHQgbGFuZz1cInRzXCIgc2V0dXA+XG5jb25zdCBtc2cgPSAnSGVsbG8gV29ybGQhJ1xuPC9zY3JpcHQ+In0=), can it be counted as a minor bug? Since it's useless in the generated code.

**Comment by edison1105** [comment]
The purpose of using `defineComponent` is to improve the support for TypeScript.

**Reply by wenfangdu** [reply]
Adding `lang="ts"` [does output `defineCompoent`](https://sfc.vuejs.org/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8aDE+e3sgbXNnIH19PC9oMT5cbjwvdGVtcGxhdGU+XG5cbjxzY3JpcHQgbGFuZz1cInRzXCIgc2V0dXA+XG5jb25zdCBtc2cgPSAnSGVsbG8gV29ybGQhJ1xuPC9zY3JpcHQ+In0=).

**Reply by edison1105** [reply]
`defineComponent`的目的是为了在开发的过程中配合IDE获得更好的ts类型提示。例如当你输入 `this.` 时会提示你定义的props,data,method等。但是编译出来的产物是不需要类型提示的，因为我们并不修改编译后的代码。

**Reply by wenfangdu** [reply]
@edison1105 多谢解答，那么加上 `lang="ts"` 会输出 `defineComponent` 算不算是小 bug 呢？因为它对编译后的代码来讲是没用的。

**Reply by edison1105** [reply]
即便是对编译后的代码没用。不能算是bug。
